pkg/api/router.go

```diff
@@ -25,198 +24,108 @@ func (s *server) setupRouting() {
+					w.Header().Set("Access-Control-Allow-Headers", "Origin, Accept, Authorization, Content-Type, X-Requested-With, Access-Control-Request-Headers, Access-Control-Request-Method, Swarm-Tag, Swarm-Pin, Swarm-Encrypt, Swarm-Index-Document, Swarm-Error-Document, Swarm-Collection, Swarm-Postage-Batch-Id, Gas-Price")
+					w.Header().Set("Access-Control-Allow-Methods", "GET, HEAD, OPTIONS, POST, PUT, DELETE")
+					w.Header().Set("Access-Control-Max-Age", "3600")
+				}
+				h.ServeHTTP(w, r)
+			})
+		},
+		s.gatewayModeForbidHeadersHandler,
+	)
 
-	router.NotFoundHandler = http.HandlerFunc(jsonhttp.NotFoundHandler)
+	r.NotFound(jsonhttp.NotFoundHandler)
+	r.MethodNotAllowed(jsonhttp.MethodNotAllowedHandler)
 
-	router.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+	r.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
 		fmt.Fprintln(w, "Ethereum Swarm Bee")
 	})
-
-	router.HandleFunc("/robots.txt", func(w http.ResponseWriter, r *http.Request) {
+	r.HandleFunc("/robots.txt", func(w http.ResponseWriter, r *http.Request) {
 		fmt.Fprintln(w, "User-agent: *\nDisallow: /")
 	})
 
+	api := chi.NewRouter()
+
 	if s.Restricted {
-		router.Handle("/auth", jsonhttp.MethodHandler{
-			"POST": web.ChainHandlers(
-				s.newTracingHandler("auth"),
-				jsonhttp.NewMaxBodyBytesHandler(512),
-				web.FinalHandlerFunc(s.authHandler),
-			),
-		})
-		router.Handle("/refresh", jsonhttp.MethodHandler{
-			"POST": web.ChainHandlers(
-				s.newTracingHandler("auth"),
-				jsonhttp.NewMaxBodyBytesHandler(512),
-				web.FinalHandlerFunc(s.refreshHandler),
-			),
-		})
+		r.With(s.newTracingHandler("auth"), jsonhttp.NewMaxBodyBytesHandler(512)).Post("/auth", s.authHandler)
+		r.With(s.newTracingHandler("auth"), jsonhttp.NewMaxBodyBytesHandler(512)).Post("/refresh", s.refreshHandler)
+		api.Use(auth.PermissionCheckHandler(s.auth))
 	}
 
-	handle("/bytes", jsonhttp.MethodHandler{
-		"POST": web.ChainHandlers(
-			s.contentLengthMetricMiddleware(),
-			s.newTracingHandler("bytes-upload"),
-			web.FinalHandlerFunc(s.bytesUploadHandler),
-		),
-	})
-	handle("/bytes/{address}", jsonhttp.MethodHandler{
-		"GET": web.ChainHandlers(
-			s.contentLengthMetricMiddleware(),
-			s.newTracingHandler("bytes-download"),
-			web.FinalHandlerFunc(s.bytesGetHandler),
-		),
+	api.Route("/bytes", func(r chi.Router) {
+		r.With(s.contentLengthMetricMiddleware(), s.newTracingHandler("bytes-upload")).Post("/", s.bytesUploadHandler)
+		r.With(s.contentLengthMetricMiddleware(), s.newTracingHandler("bytes-download")).Get("/{address}", s.bytesGetHandler)
 	})
 
-	handle("/chunks", jsonhttp.MethodHandler{
-		"POST": web.ChainHandlers(
-			jsonhttp.NewMaxBodyBytesHandler(swarm.ChunkWithSpanSize),
-			web.FinalHandlerFunc(s.chunkUploadHandler),
-		),
+	api.Route("/chunks", func(r chi.Router) {
+		r.With(jsonhttp.NewMaxBodyBytesHandler(swarm.ChunkWithSpanSize)).Post("/", s.chunkUploadHandler)
+		r.Get("/{addr}", s.chunkGetHandler)
```

Comment:
not sure if i really like this structure. I find it a bit confusing and harder to read

Reply:
we could inline it, instead of nesting it like this. The idea is for complex REST apis, this type of structuring simplifies things.